xcdat/climatology.py

```diff
@@ -0,0 +1,179 @@
+
+# xarray notation for accessing Datetime periods
+TimePeriod = Literal["time.month", "time.season", "time.year"]
+
+
+def climatology(ds: xr.Dataset, period: Period, weighted: bool = True) -> xr.Dataset:
+    """Calculates a Dataset's climatology cycle for a given period.
+
+    The "time" dimension is preserved for reference instead of being replaced
+    by the period.
+
+    The weighted averages account for leap years and each month having different
+    number of days.
+
+    Based on: https://xarray.pydata.org/en/stable/examples/monthly-means.html.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param period: The period of time to group data by
+    :type period: Period
+    :param weighted: Calculate weighted averages, defaults to True
+    :type weighted: bool, optional
+    :return: Climatology cycle for a given period
+    :rtype: xr.Dataset
+    """
+    if period not in PERIODS:
+        raise ValueError(
+            "Incorrect period argument passed. Period must be month, season, or year."
+        )
+
+    ds_base = ds.copy()
+    time_period: TimePeriod = f"time.{period}"  # type:ignore
+
+    if weighted:
+        weights = _calculate_weights(ds_base, time_period)
+        return _group_data(ds_base, time_period, weights)
+    else:
+        return _group_data(ds_base, time_period)
+
+
+def _calculate_weights(ds: xr.Dataset, time_period: TimePeriod) -> xr.DataArray:
+    """Calculates weights for a Dataset based on a period of time.
+
+    The weighted averages account for leap years and each month having different
+    number of days.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param time_period: The period of time for calculating weightes
+    :type time_period: TimePeriod
+    :return: The calculated weights based on the period of time
+    :rtype: xr.DataArray
+    """
+    month_length: xr.DataArray = ds.time.dt.days_in_month
+    weights: xr.DataArray = (
+        month_length.groupby(time_period) / month_length.groupby(time_period).sum()
+    )
+    _validate_weights(ds, weights, time_period)
+    return weights
+
+
+def _validate_weights(ds: xr.Dataset, weights: xr.DataArray, time_period: TimePeriod):
+    """Validate that the sum of the weights for a dataset equal 1.0.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param weights: The calculated weights based on a period of time
+    :type weights: xr.DataArray
+    :param time_period: The period of time to group data by
+    :type time_period: TimePeriod
+    """
+    expected_count = {
+        "time.month": 12,
+        "time.season": 4,
+        "time.year": np.unique(ds.time.dt.year.data).size,
+    }
+
+    expected = np.ones(expected_count[time_period])
+    np.testing.assert_allclose(weights.groupby(time_period).sum().values, expected)
+
+
+def departure(
+    ds_base: xr.Dataset,
+    ds_climatology: xr.Dataset,
+    weights: Optional[xr.DataArray] = None,
```

Comment:
`departure()` takes into account weighting for the climatology dataset.

The example shown below seems to only calculate departure for UNWEIGHTED datasets: https://github.com/lee1043/XCDAT/blob/master/compare_cdat_xarray/annual_cycle_departure.ipynb

Code snippet:
```
def annual_cycle_departure(d, d_ann_clim):    
    # https://github.com/willyhagi/climate-data-science/blob/master/Python-Scripts/annual_cycle.py
    d_ann_ano = d.groupby('time.month') - d_ann_clim
    return d_ann_ano
```

Possible Issues
* One API issue that might occur is that the user passes `weights` when `ds_climatology` was unweighted. This is where using xarray accessors are useful. You can store properties such as `is_weighted` (boolean), `weights` (xr.DataArray), and `period` (str) which are very useful attributes.

Reply:
@pochedls @lee1043 @chengzhuzhang @jasonb5 
FYI, I ran into a situation where encapsulating code into an xarray dataset accessor will be beneficial in terms of API usability. I understand there's an overhead in learning object-oriented programming for implementation, but it is small since we are simply extending existing xarray classes instead of writing classes from the ground up (like CDAT).

Short answer on the benefits of xarray accesors: Ability to cache attributes and properties on xarray objects for reusability. This reduces the amount of code that needs to be written and tested. I provided examples in the comment above.

I am looking into refactoring this entire file into an xarray accessor class after I finish adding unit tests.

Reply:
A simple numpy ndarray example of FP vs. OOP
```
# FP
arr = np.array([1,2,3])
np.shape(arr)
np.len(arr)
np.size(arr)


# OOP
arr = np.array([1,2,3])
arr.shape
arr.len
arr.size
```